### PR TITLE
Plugin extension + Gradle task

### PR DESCRIPTION
**Pull request by patxibocos**
- Adding plugin extension to allow configuration
- Created a Gradle task for the generator

### CODE REVIEW

**Comment by patxibocos** [comment]
I need to rebase this now :)

File: src/main/kotlin/com/patxi/poetimizely/plugin/PoetimizelyExtension.kt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package com.patxi.poetimizely.plugin
+
+open class PoetimizelyExtension(
```
Comment (ralphpina) [inline src/main/kotlin/com/patxi/poetimizely/plugin/PoetimizelyExtension.kt]: Nice, I learned about Extensions 👍.

File: src/main/kotlin/com/patxi/poetimizely/plugin/PoetimizelyPlugin.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        project.tasks.register("poetimize", GeneratorTask::class.java) { task: GeneratorTask ->
+            task.optimizelyProjectId = extension.optimizelyProjectId
+            task.optimizelyToken = extension.optimizelyToken
+        }
+    }
+}
+
+internal fun Project.poetimizely(): PoetimizelyExtension =
+    extensions.getByName(POETIMIZELY_EXTENSION_NAME) as? PoetimizelyExtension
+        ?: throw IllegalStateException("$POETIMIZELY_EXTENSION_NAME is not of the correct type")
```
Comment (ralphpina) [inline src/main/kotlin/com/patxi/poetimizely/plugin/PoetimizelyPlugin.kt]: This method is only being used in a test. This doesn't seem like something we would want to call in production. Should we move it to the test file and make it private?

File: src/test/kotlin/com/patxi/poetimizely/plugin/BuildScriptTest.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+                    optimizelyToken = 'TOKEN'
+                }
+            """.trimIndent()
+            )
+            GradleRunner.create()
+                .withProjectDir(projectDir.toFile())
+                .withPluginClasspath()
+                .build()
+        }
+    }
```
Comment (ralphpina) [inline src/test/kotlin/com/patxi/poetimizely/plugin/BuildScriptTest.kt]: 👍 nice. I think it's really helpful to cover the "we won't break your build" checkbox.

File: src/main/kotlin/com/patxi/poetimizely/plugin/PoetimizelyPlugin.kt (hunk 2 above)
Comment (patxibocos) [reply src/main/kotlin/com/patxi/poetimizely/plugin/PoetimizelyPlugin.kt]: Good catch, I was using it at some point in the production code, but not now.
So I am removing it in favor of doing this in the test:
```kotlin
project.extensions.getByName("poetimizely").shouldBeInstanceOf<PoetimizelyExtension>()
```

File: src/test/kotlin/com/patxi/poetimizely/plugin/BuildScriptTest.kt (hunk 3 above)
Comment (patxibocos) [reply src/test/kotlin/com/patxi/poetimizely/plugin/BuildScriptTest.kt]: I didn't know about the existence of [Gradle's TestKit](https://docs.gradle.org/current/userguide/test_kit.html) 🙂

File: src/main/kotlin/com/patxi/poetimizely/plugin/GeneratorTask.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import javax.inject.Inject
+
+open class GeneratorTask @Inject constructor(
+    private val optimizelyProjectId: Long,
+    private val optimizelyToken: String
+) : DefaultTask() {
+
+    @TaskAction
+    fun doAction() {
+        val service: ExperimentsService = buildExperimentsService(optimizelyToken)
```
[the remark this answers is not in the document]
Comment (patxibocos) [reply src/main/kotlin/com/patxi/poetimizely/plugin/GeneratorTask.kt]: By customizing are you referring to having multiple ways of authentication? 🤔 
We can add a more complex DSL on the extension and pass everything needed to the task…

[the remark this answers is not in the document]
Comment (ralphpina) [reply src/main/kotlin/com/patxi/poetimizely/plugin/GeneratorTask.kt]: I think I wasn't seeing the Extension solution, which as you said, would be to add this complexity and expose it via the DSL.

**Pull request merged**